llvm/docs/LangRef.rst

```diff
@@ -26965,10 +26966,20 @@ to facilitate alias analysis and underlying-object detection.
 Semantics:
 """"""""""
 
-The result of ``ptrmask(ptr, mask)`` is equivalent to
-``getelementptr ptr, (ptrtoint(ptr) & mask) - ptrtoint(ptr)``. Both the returned
-pointer(s) and the first argument are based on the same underlying object (for more
-information on the *based on* terminology see
+The result of ``ptrmask(%ptr, %mask)`` is equivalent to the following expansion,
+where ``iPtrIdx`` is the index type size of the pointer::
+
+    %intptr = ptrtoint ptr %ptr to iPtrIdx ; this may truncate
+    %masked = and iPtrIdx %intptr, %mask
+    %diff = sub iPtrIdx %masked, %intptr
```

Comment:
But %diff will always be negative right? %masked must be less than %intptr since it's the result of an and.

Reply:
Indeed, it works specifically because everything is using iPtrIdx arithmetic, and therefore (in two's complement) does not matter whether things are positive or negative, it's just modular arithmetic

Reply:
This definition is like Alive2's implementation. The difference is that right now we do a zero extend or trunc of the mask to match the pointer size.

Reply:
I was thinking of an example like this, with 16-bit pointers and 8-bit indexes:
```
  %ptr = ptr 0x6789
  %mask = i8 0x00
  // ptrmask(%ptr, %mask) expands to:
  %intptr = i8 0x89
  %masked = i8 0x00
  %diff = i8 0x77
  %result = getelementptr i8, ptr 0x6789, i8 0x77
          = ptr 0x6800 // not 0x6700!
```
Have I misunderstood how getelementptr works?

Reply:
getelementptr only works on the index bits of the pointer. So the result would be `0x6700`, not `0x6800`.

Reply:
OK but I don't see that documented in the semantics: https://llvm.org/docs/LangRef.html#id234

Reply:
Can you please update the documentation of the semantics of getelementptr?

Reply:
Done as part of https://github.com/llvm/llvm-project/pull/70015.